File: test/support/fake_sensor.py
import queue
import socket
import struct
import threading
import time

from netft_driver.protocol import Command, RECORD_STRUCT, REQUEST_STRUCT


class FakeNetFTSensor:
    def __init__(self, rate_hz=200.0):
        self.rate_hz = float(rate_hz)
        self._socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self._socket.bind(("127.0.0.1", 0))
        self._socket.settimeout(min(0.002, 0.25 / self.rate_hz))
        self._stop = threading.Event()
        self._enabled = threading.Event()
        self._enabled.set()
        self._thread = None
        self._streaming = False
        self._client_address = None
        self._requests = []
        self._commands_lock = threading.Lock()
        self._payloads = queue.Queue()
        self._rdt_sequence = 0
        self._ft_sequence = 1000
        self._rdt_skip = 0

    @property
    def host(self):
        return self._socket.getsockname()[0]

    @property
    def port(self):
        return self._socket.getsockname()[1]

    @property
    def commands(self):
        with self._commands_lock:
            return tuple(command for command, sample_count in self._requests)

    @property
    def requests(self):
        with self._commands_lock:
            return tuple(self._requests)

    def start(self):
        self._thread = threading.Thread(
            target=self._run, name="fake-netft", daemon=True
        )
        self._thread.start()
        return self

    def close(self):
        self._stop.set()
        self._socket.close()
        if self._thread is not None:
            self._thread.join(timeout=1.0)

    def __enter__(self):
        return self.start()

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

    def pause(self):
        self._enabled.clear()

    def resume(self):
        self._enabled.set()

    def queue_payload(self, payload):
        self._payloads.put(bytes(payload))

    def send_payload_now(self, payload):
        if self._client_address is None:
            raise RuntimeError("fake sensor has no streaming client")
        self._socket.sendto(bytes(payload), self._client_address)

    def queue_record(
        self,
        rdt_sequence,
        status=0,
        axes=(100, -200, 300, 10, -20, 30),
        ft_sequence=None,
    ):
        if ft_sequence is None:
            ft_sequence = self._ft_sequence
        payload = RECORD_STRUCT.pack(
            rdt_sequence & 0xFFFFFFFF,
            ft_sequence & 0xFFFFFFFF,
            status & 0xFFFFFFFF,
            *axes
        )
        self._ft_sequence = (ft_sequence + 4) & 0xFFFFFFFF
        self.queue_payload(payload)

    def skip_rdt(self, count):
        self._rdt_skip += int(count)

    def wait_for_command(self, command, count=1, timeout=1.0):
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            if self.commands.count(command) >= count:
                return True
            time.sleep(0.005)
        return False

    def _record_request(self, command, sample_count):
        with self._commands_lock:
            self._requests.append((command, sample_count))

    def _handle_request(self, payload, address):
        if len(payload) != REQUEST_STRUCT.size:
            return
        header, command_value, sample_count = REQUEST_STRUCT.unpack(payload)
        if header != 0x1234:
            return
        try:
            command = Command(command_value)
        except ValueError:
            return
        self._record_request(command, sample_count)
        if command in (Command.START_REALTIME, Command.START_BUFFERED):
            self._client_address = address
            self._streaming = True
            self._rdt_sequence = 0
        elif command in (Command.STOP_STREAMING, Command.SET_SOFTWARE_BIAS):
            self._streaming = False

    def _next_payload(self):
        try:
            payload = self._payloads.get_nowait()
            if len(payload) == RECORD_STRUCT.size:
                self._rdt_sequence = struct.unpack(">I", payload[:4])[0]
            return payload
        except queue.Empty:
            self._rdt_sequence = (
                self._rdt_sequence + 1 + self._rdt_skip
            ) & 0xFFFFFFFF
            self._rdt_skip = 0
            payload = RECORD_STRUCT.pack(
                self._rdt_sequence,
                self._ft_sequence,
                0,
                100,
                -200,
                300,
                10,
                -20,
                30,
            )
            self._ft_sequence = (self._ft_sequence + 4) & 0xFFFFFFFF
            return payload

    def _run(self):
        interval = 1.0 / self.rate_hz
        next_send = time.monotonic()
        while not self._stop.is_set():
            try:
                payload, address = self._socket.recvfrom(64)
                self._handle_request(payload, address)
            except socket.timeout:
                pass
            except OSError:
                break
            now = time.monotonic()
            if (
                self._streaming
                and self._enabled.is_set()
                and self._client_address is not None
                and now >= next_send
            ):
                try:
                    self._socket.sendto(self._next_payload(), self._client_address)
                except OSError:
                    break
                next_send = max(next_send + interval, now)

File: test/cpp/support/fake_sensor.cpp
#include "fake_sensor.hpp"

#include <arpa/inet.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <unistd.h>

#include <array>
#include <algorithm>
#include <cstring>
#include <stdexcept>

namespace netft_driver::test {
namespace {
void put32(std::vector<std::uint8_t> & bytes, std::size_t offset, std::uint32_t value)
{
  for (unsigned i = 0; i < 4; ++i) bytes[offset + i] = static_cast<std::uint8_t>(value >> (24 - 8 * i));
}
}

FakeSensor::FakeSensor(const double rate_hz) : rate_hz_(rate_hz)
{
  socket_ = ::socket(AF_INET, SOCK_DGRAM, 0);
  if (socket_ < 0) throw std::runtime_error{"fake sensor socket failed"};
  (void)::fcntl(socket_, F_SETFL, ::fcntl(socket_, F_GETFL, 0) | O_NONBLOCK);
  sockaddr_in address{};
  address.sin_family = AF_INET;
  address.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
  if (::bind(socket_, reinterpret_cast<sockaddr *>(&address), sizeof(address)) != 0) { ::close(socket_); socket_ = -1; throw std::runtime_error{"fake sensor bind failed"}; }
  socklen_t size = sizeof(address);
  ::getsockname(socket_, reinterpret_cast<sockaddr *>(&address), &size);
  port_ = ntohs(address.sin_port);
  worker_ = std::thread(&FakeSensor::run, this);
}
FakeSensor::~FakeSensor() { stopping_ = true; ::shutdown(socket_, SHUT_RDWR); if (worker_.joinable()) worker_.join(); ::close(socket_); }
void FakeSensor::pause() noexcept { enabled_ = false; }
void FakeSensor::resume() noexcept { enabled_ = true; }
void FakeSensor::queue_payload(std::vector<std::uint8_t> payload) { std::lock_guard<std::mutex> lock(mutex_); payloads_.push_back(std::move(payload)); }
void FakeSensor::send_payload_now(std::vector<std::uint8_t> payload)
{
  ::sockaddr_in peer{};
  { std::lock_guard<std::mutex> lock(mutex_); if (!has_client_) throw std::runtime_error{"fake sensor has no client"}; peer = client_; }
  (void)::sendto(socket_, payload.data(), payload.size(), 0, reinterpret_cast<sockaddr *>(&peer), sizeof(peer));
}
void FakeSensor::queue_record(
  std::uint32_t rdt_sequence, std::uint32_t status, std::uint32_t ft_sequence,
  const std::array<std::int32_t, 6> axes)
{
  std::lock_guard<std::mutex> lock(mutex_);
  if (ft_sequence == 0) ft_sequence = ft_;
  std::vector<std::uint8_t> data(36); put32(data, 0, rdt_sequence); put32(data, 4, ft_sequence); put32(data, 8, status);
  for (std::size_t index = 0; index < axes.size(); ++index) {
    put32(data, 12 + 4 * index, static_cast<std::uint32_t>(axes[index]));
  }
  ft_ = ft_sequence + 4; payloads_.push_back(std::move(data));
}
void FakeSensor::skip_rdt(unsigned count) noexcept { std::lock_guard<std::mutex> lock(mutex_); skip_ += count; }
bool FakeSensor::wait_for_command(Command command, unsigned count, std::chrono::milliseconds timeout) const
{
  const auto until = std::chrono::steady_clock::now() + timeout;
  while (std::chrono::steady_clock::now() < until) { const auto seen = commands(); if (std::count(seen.begin(), seen.end(), command) >= static_cast<long>(count)) return true; std::this_thread::sleep_for(std::chrono::milliseconds{2}); }
  return false;
}
std::vector<Command> FakeSensor::commands() const { std::lock_guard<std::mutex> lock(mutex_); return commands_; }
std::vector<FakeSensor::CommandEvent> FakeSensor::command_events() const { std::lock_guard<std::mutex> lock(mutex_); return command_events_; }
void FakeSensor::record_command(const std::uint8_t * data, std::size_t size, const ::sockaddr_in & peer)
{
  if (size != 8 || data[0] != 0x12 || data[1] != 0x34) return;
  const auto command = static_cast<Command>((static_cast<std::uint16_t>(data[2]) << 8) | data[3]);
  { std::lock_guard<std::mutex> lock(mutex_); commands_.push_back(command); command_events_.push_back({command, std::chrono::steady_clock::now()}); if (command == Command::StartRealtime) { client_ = peer; has_client_ = true; rdt_ = 0; } }
  if (command == Command::StartRealtime) streaming_ = true;
  if (command == Command::StopStreaming || command == Command::SetSoftwareBias) streaming_ = false;
}
void FakeSensor::send_next()
{
  if (!streaming_ || !enabled_) return;
  std::vector<std::uint8_t> data;
  ::sockaddr_in peer{};
  { std::lock_guard<std::mutex> lock(mutex_); if (!has_client_) return; peer = client_; if (!payloads_.empty()) { data = std::move(payloads_.front()); payloads_.erase(payloads_.begin()); }
    if (data.empty()) { data.resize(36); rdt_ += 1 + skip_; skip_ = 0; put32(data, 0, rdt_); put32(data, 4, ft_); ft_ += 4; put32(data, 12, 100); put32(data, 16, static_cast<std::uint32_t>(-200)); put32(data, 20, 300); put32(data, 24, 10); put32(data, 28, static_cast<std::uint32_t>(-20)); put32(data, 32, 30); }
  }
  ::sendto(socket_, data.data(), data.size(), 0, reinterpret_cast<sockaddr *>(&peer), sizeof(peer));
}
void FakeSensor::run()
{
  auto next = std::chrono::steady_clock::now();
  const auto interval = std::chrono::duration_cast<std::chrono::steady_clock::duration>(std::chrono::duration<double>{1.0 / rate_hz_});
  while (!stopping_) { ::sockaddr_in peer{}; socklen_t peer_size = sizeof(peer); std::array<std::uint8_t, 64> data{}; const auto size = ::recvfrom(socket_, data.data(), data.size(), 0, reinterpret_cast<sockaddr *>(&peer), &peer_size); if (size > 0) record_command(data.data(), static_cast<std::size_t>(size), peer); const auto now = std::chrono::steady_clock::now(); if (now >= next) { send_next(); next += interval; if (next < now) next = now + interval; } std::this_thread::sleep_for(std::chrono::microseconds{100}); }
}
}  // namespace netft_driver::test
